### PR TITLE
[SMALLFIX] Minor optimization for Object UFS list performance

### PR DESCRIPTION
**Pull request by madanadit**
This PR saves a few RPCs to an object store for isDirectory and list calls.

### CODE REVIEW

File: core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java [hunk 1]
```diff
@@ -258,7 +257,11 @@ public long getModificationTimeMs(String path) throws IOException {
   @Override
   public boolean isDirectory(String path) throws IOException {
     // Root is always a folder
-    return isRoot(path) || getFolderMetadata(path) != null;
+    String keyAsFolder = convertToFolderName(stripPrefixIfPresent(path));
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java]: Do you mind putting the `isRoot` check first separately?

File: core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java [hunk 2]
```diff
@@ -544,6 +518,31 @@ protected abstract ObjectListingChunk getObjectListingChunk(String key, boolean
...
+    try {
+      ObjectListingChunk objs = getObjectListingChunk(dir, recursive);
+      // If there are, this is a folder and we can create the necessary metadata
+      if (objs != null && objs.getObjectNames() != null && objs.getObjectNames().length > 0) {
+        // If the breadcrumb exists, this is a no-op
+        mkdirsInternal(dir);
+        return objs;
+      }
+    } catch (IOException e) {
+      // Return null if an exception occurs
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java]: Why swallow the exception, are exceptions expected?

File: tests/src/test/java/alluxio/underfs/UnderStorageSystemInterfaceIntegrationTest.java [hunk 3]
```diff
@@ -351,6 +351,17 @@ public void listStatus() throws IOException {
...
 
   /**
+   * Tests if listStatus returns an empty array for an empty directory.
+   */
+  @Test
+  public void listStatusEmpty() throws IOException {
+    String testDir = PathUtils.concatPath(mUnderfsAddress, "testDir");
+    mUfs.mkdirs(testDir);
+    UnderFileStatus[] res = mUfs.listStatus(testDir);
+    Assert.assertTrue(res != null && res.length == 0);
```
Comment (calvinjia) [inline tests/src/test/java/alluxio/underfs/UnderStorageSystemInterfaceIntegrationTest.java]: Could you change this to `Assert.assertEquals(0, res.length)`?

File: core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java (hunk 2 above)
Comment (madanadit) [inline core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java]: thanks. exceptions are not expected, i improved the documentation to make this clear.

**Comment by madanadit** [comment]
@calvinjia PTAL

**Comment by calvinjia** [comment]
LGTM

**Comment by ifcharming** [comment]
LGTM, thanks for the improvement

**Pull request merged**